### ISSUE TITLE
Issue with YOLO-OBB Training: NaN Loss due to Incorrect Covariance Matrices

### ISSUE THREAD

**Issue by QuantumForgeEngineer** [ultralytics/ultralytics#10108]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

Hello,

I'm encountering an issue while training YOLOv8-OBB or YOLOv5-OBB models. Specifically, during training, both the train/box_loss and the val/box_loss occasionally become NaN after a certain epoch. After debugging the code, I've identified that the variable "t3" becomes highly negative, which provokes the NaNs. Upon further investigation, I found that the issue lies with the computation of the covariance matrices.

```python

def probiou(obb1, obb2, CIoU=False, eps=1e-7):
    """
    Calculate the prob IoU between oriented bounding boxes, https://arxiv.org/pdf/2106.06072v1.pdf.

    Args:
        obb1 (torch.Tensor): A tensor of shape (N, 5) representing ground truth obbs, with xywhr format.
        obb2 (torch.Tensor): A tensor of shape (N, 5) representing predicted obbs, with xywhr format.
        eps (float, optional): A small value to avoid division by zero. Defaults to 1e-7.

    Returns:
        (torch.Tensor): A tensor of shape (N, ) representing obb similarities.
    """
    x1, y1 = obb1[..., :2].split(1, dim=-1)
    x2, y2 = obb2[..., :2].split(1, dim=-1)
    a1, b1, c1 = _get_covariance_matrix(obb1)
    a2, b2, c2 = _get_covariance_matrix(obb2)

    t1 = (
        ((a1 + a2) * (y1 - y2).pow(2) + (b1 + b2) * (x1 - x2).pow(2)) / ((a1 + a2) * (b1 + b2) - (c1 + c2).pow(2) + eps)
    ) * 0.25
    t2 = (((c1 + c2) * (x2 - x1) * (y1 - y2)) / ((a1 + a2) * (b1 + b2) - (c1 + c2).pow(2) + eps)) * 0.5
           
    t3 = (
        ((a1 + a2) * (b1 + b2) - (c1 + c2).pow(2))
        / (4 * ((a1 * b1 - c1.pow(2)).clamp_(0) * (a2 * b2 - c2.pow(2)).clamp_(0)).sqrt() + eps)
        + eps
    ).log() * 0.5
[...]

```
According to the paper that defines this function, there don't appear to be any miscoded parts. However, upon inspecting the output of the _get_covariance_matrix(obb) function, I noticed that the computed values for "a," "b," and "c" do not satisfy the condition "a*b - c^2 > 0". This condition is crucial for certain computations in the model.

The training data consists of artificially generated rectangles and circles. You can find examples of this dataset in the following GitHub issue: [Link to GitHub Issue](https://github.com/ultralytics/ultralytics/issues/9768).

If necessary for debugging purposes, I share the code for generating the dataset: https://github.com/QuantumForgeEngineer/artificial-obb-dataset 

Looking forward to any insights or suggestions on how to address this issue.

### Environment

Ultralytics YOLOv8.1.42 🚀 Python-3.9.18 torch-2.1.1+cu118 CUDA:0 (Tesla V100-PCIE-32GB, 32501MiB)
Setup complete ✅ (32 CPUs, 376.6 GB RAM, 1581.2/7333.6 GB disk)

OS                  Linux-5.19.0-32-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.9.18
Install             git
RAM                 376.57 GB
CPU                 Intel Xeon Gold 6144 3.50GHz
CUDA                11.8

matplotlib          ✅ 3.8.2>=3.3.0
numpy               ✅ 1.24.4>=1.22.2
opencv-python       ✅ 4.8.1.78>=4.6.0
pillow              ✅ 10.1.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.10.1>=1.4.1
torch               ✅ 2.1.1+cu118>=1.8.0
torchvision         ✅ 0.16.1+cu118>=0.9.0
tqdm                ✅ 4.66.1>=4.64.0
psutil              ✅ 5.9.7
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.0.3>=1.1.4
seaborn             ✅ 0.13.0>=0.11.0

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
Hello,

Thank you for the detailed report and for your efforts in pinpointing where the NaN issue might be occurring during the YOLOv8-OBB training. It indeed sounds like the negative values for "t3" stemming from your covariance matrix calculations are leading to this. The condition "a*b - c^2 > 0" is essential for ensuring positive definiteness, which is crucial for the covariance matrix.

Given that the problem arises from the condition not being satisfied, I recommend adding a check right after calculating `a`, `b`, and `c` in the `_get_covariance_matrix(obb)` function. This way, you can ensure that the matrices computed adhere to the requisite condition. If any matrix does not satisfy this condition, you could adjust the values of `a`, `b`, and `c` minimally to satisfy "a*b - c^2 > 0".

Here's a quick example of how you could add such a check:

```python
a, b, c = _get_covariance_matrix(obb)

# Ensure positive definiteness
if a*b - c.pow(2) <= 0:
    epsilon = abs(a*b - c.pow(2)) + 0.01  # Minimum adjustment
    a += epsilon
    b += epsilon

```

This is a simplistic approach and the adjustment may need further refinement based on your dataset and model specifics. Additionally, since your training data is artificially generated, I'd suggest reviewing if the geometric properties of your rectangles and circles always produce valid oriented bounding boxes respecting the positive definiteness of covariance matrices.

Your initiative in creating a dataset and sharing this issue is much appreciated. If the problem persists or if there are further insights you'd derive from subsequent tests, feel free to update or reach out again.

Best regards! 🚀

**Comment by glenn-jocher** [comment]
@QuantumForgeEngineer thanks for sharing your attempts to solve the NaN issue in your YOLOv8-OBB training! 👍

Editing 't3' directly seems like a creative approach, though it’s concerning that it still led to NaNs down the line. This hints at potential underlying issues, maybe deeper than just this calculation. Adding checks directly to 'a', 'b', and 'c' might give more stability, as it ensures the root inputs to your computations stay valid.

Skipping examples that trigger NaN, while a quick fix, might indeed not be ideal as it could lead to the "explosion" of the box_loss you observed. This is likely because it’s avoiding a symptom rather than addressing the root cause, possibly leading to an unstable training process.

As a suggestion, consider a slight modification to the covariance matrix computation to ensure it always satisfies the positive definiteness required. Here's a succinct example that may help:

```python
def ensure_pos_def(a, b, c, epsilon=1e-6):
    while a*b - c**2 <= 0:
        a += epsilon
        b += epsilon
    return a, b, c
```

And then call this function after computing your covariance matrix components:

```python
a, b, c = ensure_pos_def(a, b, c)
```

This approach aims at maintaining a valid starting point for computations that follow. Let’s keep exploring solutions, and thanks again for your contributions and detailed feedback! 🚀

**Comment by Petros626** [comment]
@QuantumForgeEngineer could you solve the issue finally? Isn't this a bug of ultralytics then?

**Comment by glenn-jocher** [comment]
Thank you for your follow-up. This issue appears to stem from the dataset or covariance matrix calculations rather than a direct bug in Ultralytics. Ensuring valid covariance matrices (e.g., `a*b - c^2 > 0`) is critical, as training stability depends on the input data's consistency. If further investigation suggests a bug in the codebase, feel free to submit a reproducible example for review.

**Comment by Petros626** [comment]
I encountered it after a few epochs with a dataset around 18k training images. I didn't use 'amp=True', bc in other issues this was the source of the problem.

**Comment by glenn-jocher** [comment]
The suggested code to ensure positive definiteness of the covariance matrices can serve as a temporary workaround, but the root cause likely lies in data handling or label formatting issues. I'd recommend thoroughly validating your dataset and labels for consistency and correctness to prevent such errors from propagating during training.

**Comment by Petros626** [comment]
What root cause could be in data using an image size of 640x640, type `.png` with normalized [0...1] labels? My default labels are in the specified YOLO OBB format.

Label train format example:
```txt
2 0.540656 0.869783 0.517220 0.869548 0.517341 0.857362 0.540778 0.857596
```
Label val format example:
```txt
1 0.223547 0.113517 0.223496 0.049611 0.258965 0.049583 0.259016 0.113489 22.58 0.00 0
3 0.578845 0.261318 0.578091 0.297560 0.564032 0.297268 0.564786 0.261025 30.98 0.00 3
```
**Note:** the last 3 values in the validation labels, get ignored using this in `ultralytics/data/utils.py`:
```python
default code...
lb = [x[:9] for x in lb] # <- added by me
if any(len(x) > 6 for x in lb) and (not keypoint):  # is segment
default code..
```

**Comment by glenn-jocher** [comment]
The root cause could be the inconsistency between your training and validation label formats. While the training labels strictly follow the YOLO OBB format, your validation labels include extraneous values (last three entries) which may lead to issues during training or evaluation. Modifying the code to truncate these values (`lb = [x[:9] for x in lb]`) is a valid temporary approach, but ensure all labels across the dataset strictly adhere to the YOLO OBB format: `class_index x1 y1 x2 y2 x3 y3 x4 y4`. For more on OBB datasets, refer to the [Oriented Bounding Box (OBB) Datasets Documentation](https://docs.ultralytics.com/datasets/obb/).

**Comment by Petros626** [comment]
I got these additional information for a difficulty filtering approach not yet applied. I'm watching to see if NaN's problem persists during further training sessions.

**Comment by glenn-jocher** [comment]
Thank you for the update. Monitoring the training process further to observe whether NaNs persist is a good approach. If issues continue, revisiting label formatting or filtering mechanisms may help maintain stability during training. Keep us posted on your findings!
